test image CI

aGuttman opened the pull request:
Changes made within `examples/em-server-multi-tier-cronjob` to use and run server images created by new CI in `e-mission-server`.

`CHANGEME` left in `docker-compose.yml` for port number so that user/tester can think about what they need for their purposes themselves.

Results of testing shown in https://github.com/e-mission/e-mission-server/pull/875.

Requisite task to show feature https://github.com/e-mission/e-mission-docs/issues/752 working.

aGuttman committed:
test image CI

Changes made within examples/em-server-multi-tier-cronjob to use and run server images created by new CI in e-mission-server.

CHANGEME left in docker-compose.yml for port number so that user/tester can think about what they need for their purposes themselves.

shankari commented on examples/em-server-multi-tier-cronjob/analysis/start_cron.sh:
```diff
@@ -1,4 +1,4 @@
-source /clone_server.sh
+source ./.docker/setup_config.sh
```
I'm confused. 

The whole point of pre-building the image is that we don't have to perform heavy-weight operations at runtime. You have already run `setup_config` here
https://github.com/e-mission/e-mission-server/pull/875/files#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557R32

So why do we have to run it again as part of starting the cronjob aka at runtime instead of during the build?

aGuttman replied:
You're right, this is silly. The one thing I wasn't sure of was whether the config was being copied over correctly, in the right place and `setup_config.sh` does have a step about that. I wanted to see it run.

The local `Dockerfile` copies the repo's local `config` dir into the image's `/config`. Then when `setup_config.sh` runs, it copies the image's `/config` into the image's working dir at `config`. I think that this is what we want to happen, but I should just do that step without the others. For this test, it seems to run fine with or without running this step.

shankari replied:
Yes, we do want to copy local `conf` -> container `/conf` -> container app `conf`
That is all we need, no need to set up conda all over again.

shankari commented on examples/em-server-multi-tier-cronjob/analysis/start_cron.sh:
```diff
@@ -29,4 +29,4 @@ pip install devcron
 echo "Launch the cronjob"
 # while true; do sleep 30; done;
-devcron ../crontab >> /var/log/cron.console.stdinout 2>&1
+devcron ./crontab >> /var/log/cron.console.stdinout 2>&1
```
I am pretty sure that this doesn't work

You are in the e-mission-server directory at this point, and the crontab is copied to `/usr/src/app`

```
COPY crontab /usr/src/app/crontab
```

Can you indicate how you tested this?

The docker-compose output has

```
em-server-multi-tier-cronjob-analysis-server-1  | Launch the cronjob
```

But nothing indicating that the cronjob was actually invoked.

You could check the logs to see if the intake scripts were actually launched every hour
I bet they are not.
Although I'm also curious about what failed and caused you to want to change this.

aGuttman replied:
So I think this is the same issue that I commented on in https://github.com/e-mission/e-mission-server/pull/875.

There is no longer a `e-mission-server` directory, so both the server source and crontab exist in `/usr/src/app`. When I'm in the correct dir to access the e-mission files, I'm also in the correct dir the access the `crontab` file.

The compose output having `em-server-multi-tier-cronjob-analysis-server-1  | Launch the cronjob` but no output about it after that is what I was seeing when I ran the original. I'll look at the logs.

It did take me a second to realize what was going wrong when the original line `devcron ../crontab >> /var/log/cron.console.stdinout 2>&1` ran and failed. It just said exited with code 1. Like I said in the other comment, I hadn't really realized that I had changed the directory structure. Perhaps this is a reason to keep it the way it was? Create some distinction between the server files themselves and these other scripts that help run the server.

shankari replied:
Yes would be good to keep the crontab out of the server code - in other words, keep `/usr/src/app` for the app only.

aGuttman replied:
Ok. So I will keep the server code directly in `/usr/src/app` and create some `/usr/src/run_server` dir for the supplemental scripts in `webbapp` and `analysis`.

shankari replied:
or you can even put them in `.docker`
Just don't want them in the main server directory.

shankari commented on examples/em-server-multi-tier-cronjob/webapp/start_script.sh:
```diff
@@ -1,10 +1,11 @@
 #!/usr/bin/env bash
-source /clone_server.sh
+
+source ./.docker/setup_config.sh
```
Ditto, this shouldn't be needed.

aGuttman replied:
Same thing here as above.

shankari commented on examples/em-server-multi-tier-cronjob/analysis/Dockerfile:
```diff
@@ -1,5 +1,5 @@
 # python 3
-FROM emission/e-mission-server.dev.server-only:4.0.0
+FROM agutt/e-mission-server:ceo_ebike_project_2022-09-19--45-23
```
This is fine for now, but we should revert it after https://github.com/e-mission/e-mission-server/pull/875 and replace it with the e-mission-server image

aGuttman replied:
Agreed. Just need to have the e-mission-server create the new images before we can use them.

shankari reviewed (changes requested):
The biggest issue is around re-creating the conda environment at runtime.
You can also see this in the docker-compose output

```
em-server-multi-tier-cronjob-analysis-server-1  | About to start conda update, this may take some time...
em-server-multi-tier-cronjob-analysis-server-1  | Installing for version 4.8.3 and platform Linux-x86_64
em-server-multi-tier-cronjob-analysis-server-1  |   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
em-server-multi-tier-cronjob-analysis-server-1  |                                  Dload  Upload   Total   Spent    Left  Speed
em-server-multi-tier-cronjob-web-server-1       | About to start conda update, this may take some time...
em-server-multi-tier-cronjob-web-server-1       | Installing for version 4.8.3 and platform Linux-x86_64
em-server-multi-tier-cronjob-web-server-1       |   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
em-server-multi-tier-cronjob-web-server-1       |                                  Dload  Upload   Total   Spent    Left  Speed
```

aGuttman committed:
Remove conda install, file reorg

Removed the use of a script that reinstalled conda while saving the config copying. Moved where supplemental scripts get copied into.

aGuttman commented:
Changes made, committed. Runs without reinstalling conda, copies the config without another script, supplemental files are copied into `.docker`, dashboard webpage loads. Is there anything left to do here?

shankari commented on examples/em-server-multi-tier-cronjob/analysis/Dockerfile:
```diff
@@ -1,12 +1,12 @@
 # python 3
-FROM emission/e-mission-server.dev.server-only:4.0.0
+FROM agutt/e-mission-server:ceo_ebike_project_2022-09-19--45-23
```
@aGuttman you still need to fix this as well. Can you make sure that all my comments (even the ones that show a resolved, are actually resolved? Some might be just postponed to cleanup.

Otherwise looks good.

aGuttman replied:
Should I use `shankari/e-mission-server:master_2022-09-22--20-37` that was just created? A little unclear, because that also seems like a temp solution. I thought I would change it when we had new `emission/e-mission-server` images to use.

shankari replied:
@aGuttman yes, please use that for now so that we can merge this and the dashboard Dockerfile.
Then, as part of cleanup, we need to figure out how to import images properly
There are pending decisions/changes in addition to the org that we publish under.

For example, the jenkins deploy now allows us to specify the branch to use. How will that be reflected here?
Also, do we have to edit the file manually every time we push a new release to the server?

aGuttman replied:
Starting image changed to `shankari/e-mission-server:master_2022-09-22--20-37`. Builds, runs, browser gets redirected to https://www.nrel.gov/transportation/openpath.html as expected.

shankari commented on examples/em-server-multi-tier-cronjob/webapp/Dockerfile:
```diff
@@ -1,7 +1,7 @@
 # python 3
-FROM emission/e-mission-server.dev.server-only:4.0.0
+FROM agutt/e-mission-server:ceo_ebike_project_2022-09-19--45-23
```
@aGuttman ditto

aGuttman committed:
Change starting image

Change starting image from one hosted by Andrew to Shankari. Tested and works.
